packages/vitest/src/node/coverage.ts

```diff
@@ -144,50 +151,64 @@ export class BaseCoverageProvider<Options extends ResolvedCoverageOptions<'istan
     }
 
     // File outside project root with default allowExternal
-    if (this.options.allowExternal === false && !filename.startsWith(this.ctx.config.root)) {
+    if (this.options.allowExternal === false && roots.every(root => !filename.startsWith(root))) {
       this.globCache.set(filename, false)
 
       return false
     }
 
-    const options: pm.PicomatchOptions = {
-      contains: true,
-      dot: true,
-      cwd: this.ctx.config.root,
-      ignore: this.options.exclude,
-    }
-
     // By default `coverage.include` matches all files, except "coverage.exclude"
     const glob = this.options.include || '**'
 
-    const included = pm.isMatch(filename, glob, options) && existsSync(cleanUrl(filename))
+    const included = roots.some((root) => {
+      const options: pm.PicomatchOptions = {
+        contains: true,
+        dot: true,
+        cwd: root,
+        ignore: this.options.exclude,
+      }
+
+      return pm.isMatch(filename, glob, options)
+    }) && existsSync(cleanUrl(filename))
```

Comment:
Format changes to improve readability:

```suggestion
    let included = roots.some(root =>
      pm.isMatch(filename, glob, {
        contains: true,
        dot: true,
        cwd: root,
        ignore: this.options.exclude,
      }))

    included = included && existsSync(cleanUrl(filename))
```